### DISCUSSION TITLE
Could multi-level directories improve performance of put or get operations?

### DISCUSSION THREAD

**Discussion by Clara12062** [minio/minio#18713] [Q&A] [answered]
Hi!
When there are a large number of objects in a bucket, does using a multi-level directory improve the performance of put/get operations or the efficiency of scanning? 

For example, if a bucket contains 1000k objects, using `mc ls` takes a long time. Using prefix filtering can improve the speed of listing. In addition to this, could this improvement have an impact on put or get operations? Or can it accelerate the scanning cycle for the scanner?

**Comment by klauspost** [comment, marked as answer]
Yes. As [Object Organization and Planning](https://min.io/docs/minio/linux/administration/object-management.html#object-organization-and-planning) says:

> * Deployments with modest or budget-focused hardware should architect their workloads to target 10,000 objects per prefix as a baseline. Increase this target based on benchmarking and monitoring of real world workloads up to what the hardware can meaningfully handle.
> * Deployments with high-performance or enterprise-grade [hardware](https://min.io/docs/minio/linux/operations/checklists/hardware.html#deploy-minio-distributed-recommendations) can typically handle prefixes with millions of objects or more.

This is similar to [amazon recommendations](https://docs.aws.amazon.com/AmazonS3/latest/userguide/optimizing-performance.html) - that limits per-prefix operations. With MinIO the performance of each prefix scales with the number of drives - but a well distributed structure is always better.

For deeper performance related questions look into our [subscriptions](https://min.io/pricing).
